File: src/UnwindMC/Util/InstructionExtensions.cs
using NDis86;
using System;

namespace UnwindMC.Util
{
    public static class InstructionExtensions
    {
        public static ulong GetTargetAddress(this Instruction instr)
        {
            var targetOffset = instr.Offset + instr.Length;
            return (ulong)((long)targetOffset + instr.Operands[0].GetValue());
        }

        public static long GetValue(this Operand operand)
        {
            switch (operand.Size)
            {
                case 8: return operand.LValue.@sbyte;
                case 16: return operand.LValue.sword;
                case 32: return operand.LValue.sdword;
                case 64: return operand.LValue.sqword;
                default: throw new InvalidOperationException();
            }
        }

        public static long GetMemoryOffset(this Operand operand)
        {
            switch (operand.Offset)
            {
                case 0: return 0;
                case 8: return operand.LValue.@sbyte;
                case 16: return operand.LValue.sword;
                case 32: return operand.LValue.sdword;
                case 64: return operand.LValue.sqword;
                default: throw new InvalidOperationException();
            }
        }
    }
}

File: src/UnwindMC/Analysis/Asm/InstructionExtensions.fs
module InstructionExtensions

open NDis86

type Operand with
    member self.GetValue (): int64 =
        match (int)self.Size with
        | 8 -> (int64)self.LValue.sbyte
        | 16 -> (int64)self.LValue.sword
        | 32 -> (int64)self.LValue.sdword
        | 64 -> self.LValue.sqword
        | _ -> impossible

    member self.GetMemoryOffset (): int64 =
        match (int)self.Offset with
        | 0 -> 0L
        | 8 -> (int64)self.LValue.sbyte
        | 16 -> (int64)self.LValue.sword
        | 32 -> (int64)self.LValue.sdword
        | 64 -> self.LValue.sqword
        | _ -> impossible

type Instruction with
    member self.GetTargetAddress (): uint64 =
        let targetOffset = self.Offset + (uint64)self.Length
        (uint64)((int64)targetOffset + self.Operands.[0].GetValue())
